miden-lib/src/notes/mod.rs

```diff
@@ -4,96 +4,104 @@ use miden_objects::{
+    let (note_script, _) = NoteScript::new(script_ast, &assembler)?;
+    let tag: Felt = target.into();
+    let serial_num = rng.draw_word();
+    Note::new(note_script.clone(), &inputs, &assets, serial_num, sender, tag)
+}
+
+/// Generates a SWAP - swap of assets between two accounts.
+pub fn create_swap_note<R: FeltRng>(
+    sender: AccountId,
+    offered_asset: Asset,
+    requested_asset: Asset,
+    mut rng: R,
+) -> Result<(Note, Note), NoteError> {
+    let assembler = TransactionKernel::assembler();
 
-    // Include the binary version of the scripts into the source file at compile time
-    let p2id_bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/P2ID.masb"));
-    let p2idr_bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/P2IDR.masb"));
     let swap_bytes = include_bytes!(concat!(env!("OUT_DIR"), "/assets/SWAP.masb"));
 
-    let (note_script_ast, inputs): (ProgramAst, Vec<Felt>) = match script {
-        Script::P2ID { target } => (
-            ProgramAst::from_bytes(p2id_bytes).map_err(NoteError::NoteDeserializationError)?,
-            vec![target.into(), ZERO, ZERO, ZERO],
-        ),
-        Script::P2IDR { target, recall_height } => (
-            ProgramAst::from_bytes(p2idr_bytes).map_err(NoteError::NoteDeserializationError)?,
-            vec![target.into(), recall_height.into(), ZERO, ZERO],
-        ),
-        Script::SWAP { asset, serial_num } => {
-            let recipient = build_p2id_recipient(sender, serial_num)?;
-            let asset_word: Word = asset.into();
-            (
-                ProgramAst::from_bytes(swap_bytes).map_err(NoteError::NoteDeserializationError)?,
-                vec![
-                    recipient[0],
-                    recipient[1],
-                    recipient[2],
-                    recipient[3],
-                    asset_word[0],
-                    asset_word[1],
-                    asset_word[2],
-                    asset_word[3],
-                    sender.into(),
-                    ZERO,
-                    ZERO,
-                    ZERO,
-                ],
-            )
-        },
-    };
-
-    let (note_script, _) = NoteScript::new(note_script_ast, &note_assembler)?;
-
-    Note::new(note_script.clone(), &inputs, &assets, serial_num, sender, tag.unwrap_or(ZERO))
-}
+    let swap_script_ast =
+        ProgramAst::from_bytes(swap_bytes).map_err(NoteError::NoteDeserializationError)?;
 
-/// Utility function generating RECIPIENT for the P2ID note script created by the SWAP script
-fn build_p2id_recipient(target: AccountId, serial_num: Word) -> Result<Digest, NoteError> {
-    // TODO: add lazy_static initialization or compile-time optimization instead of re-generating
-    // the script hash every time we call the SWAP script
-    let assembler = TransactionKernel::assembler();
+    let requested_serial_num = rng.draw_word();
+
+    let recipient = utils::build_p2id_recipient(sender, requested_serial_num.clone())?;
```

Comment:
Would be great to adapt some consistent naming throughout this function for both outgoing and returning notes.

For example, for the outgoing note we are using `swap_` but I'd use it for everything - e.g., `swap_note_script`.

For returning note, we could use something like `repay_` - e.g., `repay_serial_num`, `repay_recipient`. I don't love `repay_` though - maybe there are better alternatives?